File: python/bathymesh/contour_extractor.py
"""Heightmap processing functionality for bathymesh."""

import logging
from dataclasses import dataclass, field
from typing import List, Union

import cv2
import numpy as np
from shapely.geometry import Polygon
from shapely.ops import unary_union
from skimage import measure

from bathymesh.config import ContourParams

logger = logging.getLogger(__name__)


@dataclass
class ContourExtractor:
    """Processes 2D heightmaps to extract contours and generate polygons."""

    config: ContourParams = field(default_factory=ContourParams)

    def extract_contour_polygons(
        self,
        heightmap: np.ndarray,
        threshold: float,
    ) -> List[Polygon]:
        """
        Extract true horizontal slice polygons at threshold, handling nested contours.

        This method correctly handles cases where heightmaps have peaks and valleys
        by creating polygons with holes where appropriate. For example, a circular
        peak will produce an outer contour (going up) and inner contour (going down),
        and the result is a ring-shaped polygon.

        Args:
            heightmap: 2D numpy array of height values
            threshold: Height threshold for contour extraction
            simplify_tolerance: Initial tolerance for Douglas-Peucker simplification
            max_segments: Maximum segments per polygon before increasing simplification
            min_area_fraction: Minimum area as fraction of largest polygon

        Returns:
            List of valid Shapely Polygon objects (potentially with holes)

        Raises:
            ValueError: If heightmap is not 2D or threshold is invalid
        """
        if heightmap.ndim != 2:
            raise ValueError(f"Heightmap must be 2D, got {heightmap.ndim}D")

        logger.debug(f"Extracting contours with OpenCV at threshold {threshold}")

        # Create binary mask
        mask = (heightmap >= threshold).astype(np.uint8)

        # Find contours with hierarchy
        # RETR_CCOMP: retrieves all contours and organizes them into 2-level hierarchy
        # - External contours at hierarchy level 0
        # - Holes at hierarchy level 1
        contours_cv, hierarchy = cv2.findContours(
            mask, cv2.RETR_CCOMP, cv2.CHAIN_APPROX_SIMPLE
        )

        if len(contours_cv) == 0:
            logger.warning(f"No contours found at threshold {threshold}")
            return []

        logger.info(f"Found {len(contours_cv)} contours with OpenCV")

        # hierarchy shape: [Next, Previous, First_Child, Parent]
        # For RETR_CCOMP:
        # - If Parent == -1, it's an external contour
        # - If Parent >= 0, it's a hole of that parent
        hierarchy = hierarchy[0]  # Remove extra dimension

        try:
            contours = measure.find_contours(mask.astype(float), 0.5)
        except Exception as e:
            raise ValueError(f"Failed to extract contours: {e}") from e

        # Convert contours to polygons
        all_raw_polygons = []
        for i, contour in enumerate(contours):
            try:
                polygon = self._contour_to_polygon(contour)
                if (
                    polygon
                    and polygon.is_valid
                    and polygon.area > self.config.min_polygon_area
                ):
                    all_raw_polygons.append(polygon)
                    logger.debug(f"Added raw polygon {i} with area {polygon.area:.3f}")
            except Exception as e:
                logger.warning(f"Failed to process contour {i}: {e}")
                continue
        # Build polygons with holes
        result_polygons = []
        processed = set()

        for i, contour_cv in enumerate(contours_cv):
            if i in processed:
                continue

            # Check if this is an external contour (not a hole)
            parent_idx = hierarchy[i][3]
            if parent_idx != -1:
                # This is a hole, will be processed with its parent
                continue

            # Convert external contour to polygon
            try:
                # OpenCV contours: shape (n, 1, 2), need to reshape
                coords = contour_cv.squeeze()
                if len(coords) < 3:
                    continue

                # Flip y-axis to match numpy convention
                coords = coords[:, [0, 1]]  # x, y order

                exterior_poly = Polygon(coords)

                if (
                    not exterior_poly.is_valid
                    or exterior_poly.area < self.config.min_polygon_area
                ):
                    continue

                # Find holes (children of this contour)
                holes = []
                first_child = hierarchy[i][2]

                if first_child != -1:
                    # Process child and its siblings
                    child_idx = first_child
                    while child_idx != -1:
                        hole_contour = contours_cv[child_idx].squeeze()
                        if len(hole_contour) >= 3:
                            hole_coords = hole_contour[:, [0, 1]]
                            holes.append(hole_coords)
                            processed.add(child_idx)

                        # Move to next sibling
                        child_idx = hierarchy[child_idx][0]

                # Create polygon with holes
                if holes:
                    try:
                        poly_with_holes = Polygon(coords, holes=holes)
                        result_polygons.append(poly_with_holes)
                        logger.debug(f"Created polygon {i} with {len(holes)} holes")
                    except Exception as e:
                        logger.warning(f"Failed to create polygon with holes: {e}")
                        result_polygons.append(exterior_poly)
                else:
                    result_polygons.append(exterior_poly)

                processed.add(i)

            except Exception as e:
                logger.warning(f"Failed to process contour {i}: {e}")
                continue

        logger.info(f"Built {len(result_polygons)} polygons with OpenCV")

        # use result polygons as invert bool cut for unary union

        full_union = unary_union(all_raw_polygons)
        holes_removed = full_union.difference(unary_union(result_polygons))

        if isinstance(holes_removed, Polygon):
            result_polygons = [holes_removed]
        else:
            result_polygons = list(holes_removed.geoms)  # type: ignore

        # Filter and simplify
        result_polygons = self._filter_and_simplify_polygons(
            result_polygons,
            self.config.simplify_tolerance,
            self.config.max_segments,
            self.config.min_area_fraction,
        )

        logger.info(
            f"Final polygon count after filtering and simplification: {len(result_polygons)}"
        )
        for poly in result_polygons:
            if not poly.is_valid:
                logger.warning("Invalid polygon found after processing")

        # Remove interior geometries if they occupy >= 95% of total area
        cleaned_polygons = []
        for poly in result_polygons:
            if len(poly.interiors) > 0:
                # Calculate total hole area
                hole_area = sum(Polygon(interior).area for interior in poly.interiors)
                total_area = poly.area

                # If holes take up 95% or more, create polygon without holes
                if hole_area >= 0.95 * total_area:
                    cleaned_poly = Polygon(poly.exterior.coords)
                    cleaned_polygons.append(cleaned_poly)
                    logger.warning(
                        f"Removed interiors from polygon (hole area {hole_area:.2f} >= 95% of {total_area:.2f})"
                    )
                else:
                    cleaned_polygons.append(poly)
            else:
                cleaned_polygons.append(poly)

        result_polygons = cleaned_polygons

        return result_polygons

    def _filter_and_simplify_polygons(
        self,
        polygons: List[Polygon],
        simplify_tolerance: float,
        max_segments: int,
        min_area_fraction: float,
    ) -> List[Polygon]:
        """
        Filter and simplify polygons based on area and segment count.

        Args:
            polygons: List of polygons to filter and simplify
            simplify_tolerance: Initial tolerance for simplification
            max_segments: Maximum segments per polygon
            min_area_fraction: Minimum area as fraction of largest polygon

        Returns:
            Filtered and simplified list of polygons
        """
        if not polygons:
            return []

        # Filter by area
        max_area = max(p.area for p in polygons)
        area_threshold = max_area * min_area_fraction
        result = [p for p in polygons if p.area >= area_threshold]
        logger.info(
            f"Filtered to {len(result)} polygons with area >= {area_threshold:.3f}"
        )

        if not result:
            return []

        # Simplify if needed
        tolerance = simplify_tolerance
        while result:
            max_seg = max(len(p.exterior.coords) for p in result)
            if max_seg <= max_segments:
                break
            tolerance += 0.1
            result = [p.simplify(tolerance, preserve_topology=True) for p in result]
            logger.info(
                f"Simplified with tolerance {tolerance:.3f}, max segments now {max_seg}"
            )

        return result  # type: ignore

    def _contour_to_polygon(self, contour: np.ndarray) -> Union[Polygon, None]:
        """
        Convert contour to Shapely polygon.

        Args:
            contour: Contour coordinates from skimage

        Returns:
            Shapely Polygon or None if conversion fails
        """
        if len(contour) < 3:
            return None

        # Flip axis to get (x, y) instead of (row, col)
        coords = np.flip(contour, axis=1)

        try:
            return Polygon(coords)
        except Exception:
            return None

File: rust-implementation/src/core/contour_extractor.rs
use crate::config::ContourParams;
use anyhow::Result;
use geo::algorithm::bool_ops::BooleanOps;
use geo::algorithm::simplify::Simplify;
use geo::{Area, CoordsIter, MultiPolygon};
use geo::{Coord, LineString, Polygon};
use opencv::{
    core::{Mat, Point, Vector},
    imgproc::{self, CHAIN_APPROX_SIMPLE, RETR_CCOMP},
    prelude::*,
};

/// Represents a snapshot of contours at a specific threshold
#[derive(Debug, Clone)]
pub struct ThresholdContours {
    pub threshold: f64,
    pub contours: Vec<Polygon<f64>>,
}

pub struct ContourExtractor {
    params: ContourParams,
}

impl ContourExtractor {
    pub fn new(params: ContourParams) -> Self {
        Self { params }
    }

    /// Extract contours from a heightmap at a specific threshold
    ///
    /// This matches the Python implementation exactly, handling nested contours
    /// and creating polygons with holes where appropriate.
    pub fn extract_contours(
        &self,
        heightmap: &[f64],
        width: usize,
        height: usize,
        threshold: f64,
    ) -> Result<ThresholdContours> {
        println!("Extracting contours at threshold {:.2}", threshold);

        // Create binary mask where values >= threshold
        let mut mask_u8 = vec![0u8; width * height];
        for (i, &value) in heightmap.iter().enumerate() {
            if !value.is_nan() && value >= threshold {
                mask_u8[i] = 1;
            }
        }

        // Convert to OpenCV Mat
        // Reshape the flat slice into rows for from_slice_2d
        let rows: Vec<&[u8]> = mask_u8.chunks(width).collect();
        let mask_mat = Mat::from_slice_2d(&rows)?;

        // Find contours with hierarchy using OpenCV (matches Python's cv2.findContours)
        let mut contours_cv: Vector<Vector<Point>> = Vector::new();
        let mut hierarchy = Mat::default();

        imgproc::find_contours_with_hierarchy(
            &mask_mat,
            &mut contours_cv,
            &mut hierarchy,
            RETR_CCOMP, // 2-level hierarchy: external contours and holes
            CHAIN_APPROX_SIMPLE,
            opencv::core::Point::new(0, 0),
        )?;

        if contours_cv.is_empty() {
            println!("No contours found at threshold {:.2}", threshold);
            return Ok(ThresholdContours {
                threshold,
                contours: vec![],
            });
        }

        println!("Found {} contours with OpenCV", contours_cv.len());

        // Extract all raw polygons using skimage-like approach (contour_generation crate)
        let mut mask_f64 = vec![0.0; width * height];
        for (i, &value) in mask_u8.iter().enumerate() {
            mask_f64[i] = value as f64;
        }

        let contour_builder = contour::ContourBuilder::new(width, height, false);
        let contours_raw = contour_builder.contours(&mask_f64, &[0.5])?;

        let mut all_raw_polygons = Vec::new();
        for band in contours_raw {
            let geometry = band.geometry();
            for polygon in geometry.0.iter() {
                let coords: Vec<Coord<f64>> = polygon
                    .exterior()
                    .0
                    .iter()
                    .map(|p| Coord { x: p.x, y: p.y })
                    .collect();

                if coords.len() >= 3 {
                    let line_string = LineString::from(coords);
                    let geo_polygon = Polygon::new(line_string, vec![]);

                    if geo_polygon.unsigned_area() > self.params.min_polygon_area {
                        all_raw_polygons.push(geo_polygon);
                    }
                }
            }
        }

        // Build polygons with holes using hierarchy from OpenCV
        let mut result_polygons = Vec::new();
        let mut processed = std::collections::HashSet::new();

        for i in 0..contours_cv.len() {
            if processed.contains(&i) {
                continue;
            }

            // Get hierarchy info: [Next, Previous, First_Child, Parent]
            let hier = *hierarchy.at_2d::<opencv::core::Vec4i>(0, i as i32)?;
            let parent_idx = hier[3];

            // Skip if this is a hole (will be processed with parent)
            if parent_idx != -1 {
                continue;
            }

            // Convert external contour_generation to polygon
            let contour_generation = contours_cv.get(i)?;
            let coords = self.opencv_contour_to_coords(&contour_generation)?;

            if coords.len() < 3 {
                continue;
            }

            let exterior_poly = Polygon::new(LineString::from(coords.clone()), vec![]);

            if exterior_poly.unsigned_area() < self.params.min_polygon_area {
                continue;
            }

            // Find holes (children of this contour_generation)
            let mut holes = Vec::new();
            let first_child = hier[2];

            if first_child != -1 {
                let mut child_idx = first_child;
                while child_idx != -1 {
                    let hole_contour = contours_cv.get(child_idx as usize)?;
                    let hole_coords = self.opencv_contour_to_coords(&hole_contour)?;

                    if hole_coords.len() >= 3 {
                        holes.push(LineString::from(hole_coords));
                        processed.insert(child_idx as usize);
                    }

                    // Move to next sibling
                    let child_hier = *hierarchy.at_2d::<opencv::core::Vec4i>(0, child_idx)?;
                    child_idx = child_hier[0];
                }
            }

            // Create polygon with holes
            let final_poly = if !holes.is_empty() {
                Polygon::new(LineString::from(coords), holes)
            } else {
                exterior_poly
            };

            result_polygons.push(final_poly);
            processed.insert(i);
        }

        println!("Built {} polygons with OpenCV", result_polygons.len());

        // Skip expensive union operations for now - just use result_polygons directly
        // The union operations are O(n²) and extremely slow for large polygon counts
        // TODO: Implement spatial partitioning or cascaded union if needed
        println!(
            "Using {} result polygons directly (skipping union operations)",
            result_polygons.len()
        );

        let mut holes_removed_polys = result_polygons;

        // Filter and simplify
        holes_removed_polys = self.filter_and_simplify_polygons(
            holes_removed_polys,
            self.params.simplify_tolerance,
            self.params.max_segments,
            self.params.min_area_fraction,
        );

        println!(
            "Final polygon count after filtering and simplification: {}",
            holes_removed_polys.len()
        );

        // Remove interior geometries if they occupy >= 95% of total area
        let mut cleaned_polygons = Vec::new();
        for poly in holes_removed_polys {
            if !poly.interiors().is_empty() {
                let hole_area: f64 = poly
                    .interiors()
                    .iter()
                    .map(|ring| Polygon::new(ring.clone(), vec![]).unsigned_area())
                    .sum();
                let total_area = poly.unsigned_area();

                if hole_area >= 0.95 * total_area {
                    let cleaned_poly = Polygon::new(poly.exterior().clone(), vec![]);
                    cleaned_polygons.push(cleaned_poly);
                    println!(
                        "Removed interiors from polygon (hole area {:.2} >= 95% of {:.2})",
                        hole_area, total_area
                    );
                } else {
                    cleaned_polygons.push(poly);
                }
            } else {
                cleaned_polygons.push(poly);
            }
        }

        Ok(ThresholdContours {
            threshold,
            contours: cleaned_polygons,
        })
    }

    /// Convert OpenCV contour_generation Vector to vector of coordinates
    fn opencv_contour_to_coords(
        &self,
        contour_generation: &Vector<Point>,
    ) -> Result<Vec<Coord<f64>>> {
        let mut coords = Vec::new();

        for i in 0..contour_generation.len() {
            let point = contour_generation.get(i)?;
            coords.push(Coord {
                x: point.x as f64,
                y: point.y as f64,
            });
        }

        Ok(coords)
    }

    /// Filter and simplify polygons based on area and segment count
    fn filter_and_simplify_polygons(
        &self,
        mut polygons: Vec<Polygon<f64>>,
        simplify_tolerance: f64,
        max_segments: usize,
        min_area_fraction: f64,
    ) -> Vec<Polygon<f64>> {
        if polygons.is_empty() {
            return vec![];
        }

        // Filter by area
        let max_area = polygons
            .iter()
            .map(|p| p.unsigned_area())
            .max_by(|a, b| a.partial_cmp(b).unwrap())
            .unwrap_or(0.0);

        let area_threshold = max_area * min_area_fraction;
        polygons.retain(|p| p.unsigned_area() >= area_threshold);

        println!(
            "Filtered to {} polygons with area >= {:.3}",
            polygons.len(),
            area_threshold
        );

        if polygons.is_empty() {
            return vec![];
        }

        // Simplify if needed
        let mut tolerance = simplify_tolerance;
        loop {
            let max_seg = polygons
                .iter()
                .map(|p| p.exterior().coords_count())
                .max()
                .unwrap_or(0);

            if max_seg <= max_segments {
                break;
            }

            tolerance += 0.1;
            polygons = polygons.iter().map(|p| p.simplify(&tolerance)).collect();

            println!(
                "Simplified with tolerance {:.3}, max segments now {}",
                tolerance, max_seg
            );
        }

        polygons
    }

    /// Extract contours at multiple thresholds
    pub fn extract_multi_threshold(
        &self,
        heightmap: &[f64],
        width: usize,
        height: usize,
        thresholds: &[f64],
    ) -> Result<Vec<ThresholdContours>> {
        let mut snapshots = Vec::new();

        for &threshold in thresholds {
            let snapshot = self.extract_contours(heightmap, width, height, threshold)?;
            snapshots.push(snapshot);
        }

        Ok(snapshots)
    }
}
